File: converter.py
import os
import shutil
import subprocess
import threading
import time
import queue
from db import get_connection
from logger import get_logger

log = get_logger("converter")

FFMPEG_BIN = "ffmpeg"


def find_ffmpeg():
    """Locate ffmpeg binary, checking common install paths on Windows."""
    for candidate in [
        FFMPEG_BIN,
        r"C:\ProgramData\chocolatey\bin\ffmpeg.exe",
        os.path.expandvars(r"%LOCALAPPDATA%\Microsoft\WinGet\Links\ffmpeg.exe"),
    ]:
        try:
            result = subprocess.run(
                [candidate, "-version"],
                capture_output=True, text=True, timeout=5,
            )
            if result.returncode == 0:
                return candidate
        except (FileNotFoundError, subprocess.TimeoutExpired):
            continue
    return None


def convert_to_aiff(source_path, output_dir, ffmpeg_bin=None):
    """Convert an audio file to AIFF using FFmpeg.

    Returns (output_path, error_msg). On success error_msg is empty.
    If the source is already AIFF, copies it instead.
    """
    if not ffmpeg_bin:
        ffmpeg_bin = find_ffmpeg()
    if not ffmpeg_bin:
        return "", "ffmpeg not found"

    if not os.path.isfile(source_path):
        return "", f"source file not found: {source_path}"

    os.makedirs(output_dir, exist_ok=True)

    base = os.path.splitext(os.path.basename(source_path))[0]
    output_path = os.path.join(output_dir, base + ".aiff")

    # Avoid overwriting -- add a counter suffix if needed
    counter = 1
    while os.path.exists(output_path):
        output_path = os.path.join(output_dir, f"{base}_{counter}.aiff")
        counter += 1

    src_ext = os.path.splitext(source_path)[1].lower()

    # Already AIFF -- just copy
    if src_ext in (".aiff", ".aif"):
        import shutil
        try:
            shutil.copy2(source_path, output_path)
            return output_path, ""
        except Exception as e:
            return "", str(e)

    # FFmpeg conversion: lossless PCM into AIFF container
    cmd = [
        ffmpeg_bin,
        "-i", source_path,
        "-c:a", "pcm_s16be",   # 16-bit big-endian PCM (standard AIFF)
        "-f", "aiff",
        "-y",
        output_path,
    ]

    try:
        result = subprocess.run(
            cmd, capture_output=True, text=True, timeout=600,
        )
        if result.returncode != 0:
            stderr = result.stderr[-500:] if result.stderr else "unknown error"
            if os.path.exists(output_path):
                os.remove(output_path)
            return "", stderr
        return output_path, ""
    except subprocess.TimeoutExpired:
        if os.path.exists(output_path):
            os.remove(output_path)
        return "", "conversion timed out (10 min limit)"
    except Exception as e:
        return "", str(e)


class ConversionWorker:
    """Background thread that processes conversion jobs from a queue."""

    def __init__(self):
        self._queue = queue.Queue()
        self._thread = None
        self._running = False
        self._ffmpeg_bin = None

    @property
    def is_running(self):
        return self._running and self._thread and self._thread.is_alive()

    @property
    def queue_size(self):
        return self._queue.qsize()

    def start(self):
        if self.is_running:
            return
        self._ffmpeg_bin = find_ffmpeg()
        if not self._ffmpeg_bin:
            log.error("ffmpeg not found on this system")
            raise RuntimeError("ffmpeg not found on this system")
        log.info("Conversion worker started (ffmpeg=%s)", self._ffmpeg_bin)
        self._running = True
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()

    def stop(self):
        self._running = False
        self._queue.put(None)

    def enqueue(self, download_id, source_path, source_ext, output_dir):
        """Add a conversion job to the queue. Creates the DB record."""
        conn = get_connection()
        try:
            existing = conn.execute(
                "SELECT id FROM conversions WHERE source_path = ? AND status IN ('pending','converting','done')",
                (source_path,),
            ).fetchone()
            if existing:
                log.debug("Conversion already exists (id=%d) for %s, skipping", existing["id"], os.path.basename(source_path))
                return existing["id"]

            cur = conn.execute(
                "INSERT INTO conversions (download_id, source_path, source_ext, status) VALUES (?, ?, ?, 'pending')",
                (download_id, source_path, source_ext),
            )
            conv_id = cur.lastrowid
            conn.commit()
        finally:
            conn.close()

        log.info("Queued conversion #%d: %s -> aiff (output: %s)", conv_id, os.path.basename(source_path), output_dir)
        self._queue.put((conv_id, source_path, output_dir))
        return conv_id

    def _run(self):
        while self._running:
            try:
                job = self._queue.get(timeout=2)
            except queue.Empty:
                continue

            if job is None:
                break

            conv_id, source_path, output_dir = job
            self._process(conv_id, source_path, output_dir)

    def _process(self, conv_id, source_path, output_dir):
        fname = os.path.basename(source_path)
        log.info("Converting #%d: %s", conv_id, fname)
        conn = get_connection()
        try:
            conn.execute(
                "UPDATE conversions SET status='converting', started_at=CURRENT_TIMESTAMP WHERE id=?",
                (conv_id,),
            )
            conn.commit()

            start_time = time.time()
            output_path, error = convert_to_aiff(source_path, output_dir, self._ffmpeg_bin)
            elapsed = round(time.time() - start_time, 1)

            if error:
                log.error("Conversion #%d FAILED (%.1fs): %s -- %s", conv_id, elapsed, fname, error[:200])
                conn.execute(
                    "UPDATE conversions SET status='failed', error_msg=?, finished_at=CURRENT_TIMESTAMP WHERE id=?",
                    (error[:1000], conv_id),
                )
            else:
                try:
                    size_mb = round(os.path.getsize(output_path) / (1024 * 1024), 2)
                except OSError:
                    size_mb = 0
                log.info("Conversion #%d DONE (%.1fs): %s -> %s (%.1f MB)", conv_id, elapsed, fname, os.path.basename(output_path), size_mb)
                conn.execute(
                    "UPDATE conversions SET status='done', output_path=?, size_mb=?, finished_at=CURRENT_TIMESTAMP WHERE id=?",
                    (output_path, size_mb, conv_id),
                )

                # Move source file to _converted subfolder
                try:
                    source_dir = os.path.dirname(source_path)
                    converted_dir = os.path.join(source_dir, "_converted")
                    os.makedirs(converted_dir, exist_ok=True)
                    dest = os.path.join(converted_dir, fname)
                    if os.path.isfile(source_path):
                        shutil.move(source_path, dest)
                        conn.execute(
                            "UPDATE downloads SET filepath = ? WHERE filepath = ?",
                            (dest, source_path),
                        )
                        conn.execute(
                            "UPDATE conversions SET source_path = ? WHERE id = ?",
                            (dest, conv_id),
                        )
                        log.info("Moved source to _converted: %s", fname)
                except Exception as move_err:
                    log.warning("Could not move source to _converted: %s", move_err)

            conn.commit()
        except Exception as e:
            log.error("Conversion #%d EXCEPTION: %s", conv_id, e, exc_info=True)
            try:
                conn.execute(
                    "UPDATE conversions SET status='failed', error_msg=?, finished_at=CURRENT_TIMESTAMP WHERE id=?",
                    (str(e)[:1000], conv_id),
                )
                conn.commit()
            except Exception:
                pass
        finally:
            conn.close()

File: src-tauri/src/converter.rs
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};

use rusqlite::{Connection, params};
use tauri::{AppHandle, Emitter};
use tokio::sync::mpsc;

use crate::error::AppError;

/// A single conversion job to be processed by the worker.
pub struct ConvJob {
    pub conv_id: i64,
    pub source_path: PathBuf,
    pub output_dir: PathBuf,
}

/// Background worker that processes audio-to-AIFF conversion jobs.
pub struct ConversionWorker {
    pub sender: mpsc::Sender<ConvJob>,
    pub queue_size: Arc<AtomicUsize>,
    pub running: Arc<AtomicBool>,
}

impl ConversionWorker {
    /// Spawn the background conversion worker.
    /// Returns immediately; the worker processes jobs from the channel in a tokio task.
    pub fn start(db: Arc<Mutex<Connection>>, app: AppHandle) -> Self {
        let (tx, mut rx) = mpsc::channel::<ConvJob>(256);
        let queue_size = Arc::new(AtomicUsize::new(0));
        let running = Arc::new(AtomicBool::new(true));

        let qs = queue_size.clone();
        let r = running.clone();

        tokio::spawn(async move {
            let ffmpeg_bin = find_ffmpeg().await;
            if ffmpeg_bin.is_none() {
                tracing::error!("ffmpeg not found on this system");
            }

            while r.load(Ordering::Relaxed) {
                match rx.recv().await {
                    Some(job) => {
                        process_job(&db, &app, &ffmpeg_bin, job).await;
                        qs.fetch_sub(1, Ordering::Relaxed);
                    }
                    None => break,
                }
            }

            tracing::info!("Conversion worker shut down");
        });

        ConversionWorker {
            sender: tx,
            queue_size,
            running,
        }
    }

    /// Enqueue a conversion job. Creates the DB record and sends the job to the worker.
    /// Returns the conversion ID (existing or newly created).
    pub fn enqueue(
        &self,
        db: &Arc<Mutex<Connection>>,
        download_id: i64,
        source_path: PathBuf,
        source_ext: String,
        output_dir: PathBuf,
    ) -> Result<i64, AppError> {
        let conn = db.lock().unwrap();
        let source_str = source_path.to_string_lossy().to_string();

        // Check for existing pending/converting/done conversion (dedup)
        let existing: Option<i64> = conn
            .query_row(
                "SELECT id FROM conversions WHERE source_path = ? AND status IN ('pending','converting','done')",
                params![source_str],
                |row| row.get(0),
            )
            .ok();

        if let Some(id) = existing {
            tracing::debug!(
                "Conversion already exists (id={}) for {}, skipping",
                id,
                source_path.file_name().unwrap_or_default().to_string_lossy()
            );
            return Ok(id);
        }

        let mut stmt = conn.prepare(
            "INSERT INTO conversions (download_id, source_path, source_ext, status) VALUES (?, ?, ?, 'pending')",
        )?;
        let conv_id = stmt.insert(params![download_id, source_str, source_ext])?;
        drop(conn);

        tracing::info!(
            "Queued conversion #{}: {} -> aiff (output: {})",
            conv_id,
            source_path.file_name().unwrap_or_default().to_string_lossy(),
            output_dir.display()
        );

        self.queue_size.fetch_add(1, Ordering::Relaxed);
        let _ = self.sender.try_send(ConvJob {
            conv_id,
            source_path,
            output_dir,
        });

        Ok(conv_id)
    }

    /// Re-enqueue an existing conversion record (for retries or startup restore).
    pub fn reenqueue(&self, conv_id: i64, source_path: PathBuf, output_dir: PathBuf) {
        self.queue_size.fetch_add(1, Ordering::Relaxed);
        let _ = self.sender.try_send(ConvJob {
            conv_id,
            source_path,
            output_dir,
        });
    }

    /// Check whether the worker task is still running.
    pub fn is_running(&self) -> bool {
        self.running.load(Ordering::Relaxed)
    }

    /// Get the number of jobs currently in the queue.
    pub fn get_queue_size(&self) -> usize {
        self.queue_size.load(Ordering::Relaxed)
    }
}

/// Process a single conversion job: update DB, run FFmpeg, move source file.
async fn process_job(
    db: &Arc<Mutex<Connection>>,
    app: &AppHandle,
    ffmpeg_bin: &Option<PathBuf>,
    job: ConvJob,
) {
    let fname = job
        .source_path
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();

    tracing::info!("Converting #{}: {}", job.conv_id, fname);

    // Mark as converting
    {
        let conn = db.lock().unwrap();
        let _ = conn.execute(
            "UPDATE conversions SET status='converting', started_at=CURRENT_TIMESTAMP WHERE id=?",
            params![job.conv_id],
        );
    }
    emit_conversion_update(app, job.conv_id, "converting");

    let (output_path, error) =
        convert_to_aiff(&job.source_path, &job.output_dir, ffmpeg_bin).await;

    if !error.is_empty() {
        tracing::error!("Conversion #{} FAILED: {} -- {}", job.conv_id, fname, &error[..error.len().min(200)]);
        let conn = db.lock().unwrap();
        let _ = conn.execute(
            "UPDATE conversions SET status='failed', error_msg=?, finished_at=CURRENT_TIMESTAMP WHERE id=?",
            params![&error[..error.len().min(1000)], job.conv_id],
        );
        emit_conversion_update(app, job.conv_id, "failed");
    } else {
        let size_mb = std::fs::metadata(&output_path)
            .map(|m| (m.len() as f64) / (1024.0 * 1024.0))
            .map(|s| (s * 100.0).round() / 100.0)
            .unwrap_or(0.0);

        let out_fname = Path::new(&output_path)
            .file_name()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string();

        tracing::info!(
            "Conversion #{} DONE: {} -> {} ({:.1} MB)",
            job.conv_id,
            fname,
            out_fname,
            size_mb
        );

        {
            let conn = db.lock().unwrap();
            let _ = conn.execute(
                "UPDATE conversions SET status='done', output_path=?, size_mb=?, finished_at=CURRENT_TIMESTAMP WHERE id=?",
                params![output_path, size_mb, job.conv_id],
            );
        }

        // Move source file to _converted/ subfolder
        if let Some(source_dir) = job.source_path.parent() {
            let converted_dir = source_dir.join("_converted");
            if std::fs::create_dir_all(&converted_dir).is_ok() {
                let dest = converted_dir.join(&fname);
                if job.source_path.is_file() {
                    match std::fs::rename(&job.source_path, &dest) {
                        Ok(()) => {
                            let conn = db.lock().unwrap();
                            let source_str = job.source_path.to_string_lossy().to_string();
                            let dest_str = dest.to_string_lossy().to_string();
                            let _ = conn.execute(
                                "UPDATE downloads SET filepath = ? WHERE filepath = ?",
                                params![dest_str, source_str],
                            );
                            let _ = conn.execute(
                                "UPDATE conversions SET source_path = ? WHERE id = ?",
                                params![dest_str, job.conv_id],
                            );
                            tracing::info!("Moved source to _converted: {}", fname);
                        }
                        Err(e) => {
                            tracing::warn!("Could not move source to _converted: {}", e);
                        }
                    }
                }
            }
        }

        emit_conversion_update(app, job.conv_id, "done");
    }
}

/// Convert an audio file to AIFF format.
/// Returns (output_path, error_msg). On success error_msg is empty.
async fn convert_to_aiff(
    source_path: &Path,
    output_dir: &Path,
    ffmpeg_bin: &Option<PathBuf>,
) -> (String, String) {
    let ffmpeg = match ffmpeg_bin {
        Some(f) => f.clone(),
        None => return (String::new(), "ffmpeg not found".to_string()),
    };

    if !source_path.is_file() {
        return (
            String::new(),
            format!("source file not found: {}", source_path.display()),
        );
    }

    if std::fs::create_dir_all(output_dir).is_err() {
        return (
            String::new(),
            format!("cannot create output directory: {}", output_dir.display()),
        );
    }

    let stem = source_path
        .file_stem()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();

    let mut output_path = output_dir.join(format!("{}.aiff", stem));

    // Avoid overwriting: add _1, _2 suffix
    let mut counter = 1u32;
    while output_path.exists() {
        output_path = output_dir.join(format!("{}_{}.aiff", stem, counter));
        counter += 1;
    }

    let src_ext = source_path
        .extension()
        .unwrap_or_default()
        .to_string_lossy()
        .to_lowercase();

    // Already AIFF: just copy
    if src_ext == "aiff" || src_ext == "aif" {
        match std::fs::copy(source_path, &output_path) {
            Ok(_) => return (output_path.to_string_lossy().to_string(), String::new()),
            Err(e) => return (String::new(), e.to_string()),
        }
    }

    // FFmpeg conversion: lossless PCM into AIFF container
    let result = tokio::process::Command::new(&ffmpeg)
        .args([
            "-i",
            &source_path.to_string_lossy(),
            "-c:a",
            "pcm_s16be",
            "-f",
            "aiff",
            "-y",
            &output_path.to_string_lossy(),
        ])
        .output()
        .await;

    match result {
        Ok(output) => {
            if !output.status.success() {
                let stderr = String::from_utf8_lossy(&output.stderr);
                let truncated = if stderr.len() > 500 {
                    &stderr[stderr.len() - 500..]
                } else {
                    &stderr
                };
                let _ = std::fs::remove_file(&output_path);
                (String::new(), truncated.to_string())
            } else {
                (output_path.to_string_lossy().to_string(), String::new())
            }
        }
        Err(e) => {
            let _ = std::fs::remove_file(&output_path);
            (String::new(), e.to_string())
        }
    }
}

/// Locate the ffmpeg binary, checking common install paths on Windows.
pub async fn find_ffmpeg() -> Option<PathBuf> {
    let candidates: Vec<PathBuf> = {
        let mut c = vec![PathBuf::from("ffmpeg")];

        #[cfg(target_os = "windows")]
        {
            c.push(PathBuf::from(
                r"C:\ProgramData\chocolatey\bin\ffmpeg.exe",
            ));
            if let Ok(local) = std::env::var("LOCALAPPDATA") {
                c.push(PathBuf::from(format!(
                    r"{}\Microsoft\WinGet\Links\ffmpeg.exe",
                    local
                )));
            }
        }

        c
    };

    for candidate in candidates {
        let result = tokio::process::Command::new(&candidate)
            .arg("-version")
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .status()
            .await;

        if let Ok(status) = result {
            if status.success() {
                tracing::info!("Found ffmpeg at: {}", candidate.display());
                return Some(candidate);
            }
        }
    }

    None
}

/// Emit a conversion-update event to the frontend.
fn emit_conversion_update(app: &AppHandle, conv_id: i64, status: &str) {
    let _ = app.emit(
        "conversion-update",
        serde_json::json!({ "id": conv_id, "status": status }),
    );
}
